[FEAT] 멘토링 페이지 예약 많은 순 정렬

kimdevv opened the pull request:
## Issue Number
closed #778

## As-Is
* 정렬 기준이 최신순 하나밖에 없어서, 고객은 다른 정렬 기준을 선택할 수 없었습니다.

## To-Be
* `MentoringStatistics` 테이블의 `ReservationCount` 필드를 기준으로 내림차순 정렬하여 멘토링 페이지 조회가 가능합니다.
* 멘토링 페이지네이션 조회 로직을 리팩토링하였습니다.
* `MentoringStatistics` 테이블과 `Mentoring` 테이블의 연관 관계를 제거하였습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

kimdevv committed:
feat: 예약 수 SortKey 추가

kimdevv committed:
feat: 예약 수 cursorCondition 추가

kimdevv committed:
feat: 멘토링 페이지 조회 시 예약 개수 내림차순 기능 구현

kimdevv committed:
test: 테스트 코드에 빈 임포트

kimdevv committed:
test: 테스트 케이스 보완

kimdevv committed:
refactor: MentoringStatistics 연관 관계 제거

kimdevv committed:
refactor: update 쿼리 트랜잭션 적용

kimdevv committed:
test: 멘토링 예약 많은 순서 정렬 페이지 조회 테스트 케이스 추가

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/MentoringStatistics.java:
```diff
@@ -55,21 +57,15 @@ public class MentoringStatistics {
     @Column(name = "deleted_at")
     private LocalDateTime deletedAt;
 
-    @JoinColumn(name = "mentoring_id")
-    @OneToOne(fetch = FetchType.LAZY)
-    @MapsId()
-    private Mentoring mentoring;
-
```
`Mentoring` 테이블과의 연관 관계를 제거하였습니다. `mentoringStatistics.getMentoring()`이 사용될 경우가 절대 없을 것이라고 판단하여 연관 관계의 필요성을 크게 느끼지 못했습니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -24,87 +17,41 @@ public class CustomMentoringRepositoryImpl implements CustomMentoringRepository
+    private static final QMentoringStatistics mentoringStatistics = QMentoringStatistics.mentoringStatistics;
 
     private final JPAQueryFactory jpaQueryFactory;
+    private final MentoringPaginationHelper mentoringPaginationHelper;
```
기존에 `CustomMentoringRepositoryImpl`의 책임이 너무 거대했습니다. 그래서 `where`절과 `orderBy`절, `cursorCode`를 만드는 로직을 `MentoringPaginationHelper` 클래스로 분리시켰습니다. 이제 `CustomMentoringRepositoryImpl`는 `MentoringPaginationHelper`의 기능을 조립하여 SQL을 실행시키고 결과값을 반환시키는 역할만 담당하게 됩니다.

ppparkta replied:
p3: 작업이 쉽지 않으셨을텐데 잘 나눠주셨네요. 리뷰하는 입장에서 많이 보기 편해졌습니다!! 👍

jye326 replied:
👍

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/CustomMentoringRepositoryImpl.java:
p3: 이전에 의견 냈던 코멘트인데 반영되기 전에 닫혀서 여기에 작성합니다.

- repository interface: 구현체가 무엇이든 변하지 않고 늘 동일한 기능을 함 -> 비즈니스 로직에 속함
- 구현체: 구현하는 방식에 따라 얼마든지 변할 수 있음 -> 인프라 의존적임

따라서 인터페이스는 지금처럼 repository에 두고, 구현체는 infra에 옮기는거 어떠신가요!

kimdevv replied:
좋은 의견인 것 같습니다. 그런데 이 구현체의 패키지를 수정할 경우 `ListCrudRepository`를 확장한 다른 repository들도 infra로 옮겨야 할 것 같다는 생각이 듭니다!
이들은 일단 타입은 interface이지만 실제로는 `ListCrudRepository`를 확장하여 마치 구현체로 쓰고 있습니다. 이러한 방법이 언제든지 바뀔 수 있을 것 같아서요!(`ListCrudRepository`를 안 쓰고 다른 방식을 쓴다던가)

그래서 뭔가 대이동이 있을 것 같은 느낌이라... 추후에 다시 논의를 해봐야 할 것 같습니다.

ppparkta replied:
> 이 구현체의 패키지를 수정할 경우 ListCrudRepository를 확장한 다른 repository들도 infra로 옮겨야 할 것 같다는 생각이 듭니다!

`@Query` 사용으로 인해 인터페이스가 구현체와 비슷한 성격을 띄고 있어서 구분이 모호한 것 인정합니다! 다만 JPA를 사용하는 한 ListCrud가 Jpa나 다른 인터페이스로 교체된다고 해도 문제되지 않을 것 같아요. 그래서 인터페이스라는 성질 자체는 그대로 유지된다고 생각합니다.

그래서 전면 이동보다는, QueryDSL에 의존적인 helper/custom repository부터 인프라로 분리하는 것이 좋을 것 같습니다.

kimdevv replied:
아 이해했습니다! 레포지토리 인터페이스는 항상 동일한 결과를 예상하고 있지만 그 내부 구현은 언제든지 달라질 수 있다는 거군요. 구현체와 helper은 다른 패키지로 이동해야 할 필요성이 있어 보이네요. 일단 임시적으로 패키지 이동을 해두겠습니다! 구체적인 패키지 구조는 추후에 다같이 논의해 보면 좋을 것 같습니다!

kimdevv replied:
그런데 패키지를 수정하니 JPA 관련 오류가 나서 추후에 논의 후 바꿔야 될 것 같습니다... 일단 원상복구 시킬게요

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/MentoringStatisticsRepository.java:
```diff
@@ -5,9 +5,11 @@
 
 public interface MentoringStatisticsRepository extends ListCrudRepository<MentoringStatistics, Long> {
 
+    @Transactional
```
p3: 이거 보고 찾아보니까 Modifying 어노테이션에 트랜잭셔널이 안 붙네요...?! 중요한 이슈였는데 함께 처리해주셔서 감사합니다

ppparkta replied:
p2: 관련해서 글 찾아보다가 Modifying만으로는 영속성 컨텍스트 관리가 보장되지 않는다는 것을 발견했어요. 서비스 안에 여러 쿼리가 사용되는 경우 아래처럼 수정할 것을 권장한다고 하네요. 

혼란을 막기 위해서 `@Modifying`을 사용할 때는 batch 처리 등 특별한 경우를 제외하고 모두 옵션을 추가해주는 것 어떠신가요?
```java
@Modifying(clearAutomatically = true, flushAutomatically = true)
```

https://www.baeldung.com/spring-data-jpa-modifying-annotation

kimdevv replied:
좋은 글이네요. 두 속성을 추가함으로써 `항상 최신 상태의 레코드를 제거할 수 있고, 항상 최신 상태의 레코드를 불러오도록` 할 수 있겠군요~

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/MentoringServiceTest.java:
```diff
@@ -1015,4 +1015,14 @@ void modifyMentoringFail2() {
         }
     }
+
+    @DisplayName("멘토링을 아무 조건 및 필터링 없이 조회할 수 있다")
+    @Test
+    void findMentoringSummaryPages1() {
+        // given
+
+        // when
+
+        // then
+    }
 }
```
p2: 이 테스트는 지우는걸 깜빡하신건가요??!

kimdevv replied:
헉 감사합니다. 빠뜨렸네요!!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
p3: MentoringStatistics 테이블이 생겼으니 기존에 Mentoring 리뷰 정보(평점, 개수) 구하는 쿼리를 MentoringStatistics 테이블로 조회하도록 수정하면 훨씬 빨라지겠네요! 이번에 같이 반영해주시면 좋을 것 같아요.

ppparkta reviewed (commented):
이번에 클래스 분리를 잘 해주신 덕분에 너무 너무 이해하기 편했어요. 감사합니다! 개인적으로 팀 모두가 queryDSL에 익숙해지기까지는 지금처럼 주석을 남겨놓고, 익숙해지면 전체적으로 주석 정리하는 시간 가지면 좋을 것 같아요! 궁금한 부분 코멘트 몇 개 남겼습니다. 수고하셨어요!

kimdevv committed:
refactor: @Modifying 어노테이션에 자동 플러시 및 클리어 설정 추가

- @Modifying에 `flushAutomatically`와 `clearAutomatically` 속성 추가
- 수정성 쿼리 수행 시 영속성 컨텍스트 상태를 자동으로 동기화하도록 변경

kimdevv committed:
test: 불필요한 테스트 케이스 제거

- 테스트 코드에서 `findMentoringSummaryPages1` 메서드 삭제
- 사용되지 않는 코드 정리

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/MentoringStatistics.java:
```diff
@@ -28,6 +28,8 @@
+    @Getter
+    @Column(name = "mentoring_id")
     @Id
     private Long id;
```
p3 : 칼럼명과 변수명을 통일하기 위해 mentoring_id로 변경하는 건 어떨까요?

kimdevv replied:
그게 덜 헷갈리겠네요. 수정하겠습니다~

kimdevv committed:
refactor: 리뷰 통계 조회 로직 MentoringStatistics로 통합

- ReviewRepository와 관련 쿼리 제거
- MentoringStatisticsRepository에 통계 조회 메서드 추가
- MentoringStatistics에 평균 평점 계산 메서드 추가
- 코드 간소화를 위해 Review 및 Mentoring 관련 불필요 메서드, 테스트 케이스 제거
- 코드 정리 및 일부 로직 최적화

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/helper/MentoringPaginationHelper.java:
주석 덕분에 읽기 수월하네요👍

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/CustomMentoringRepositoryImpl.java:
```diff
@@ -24,87 +17,41 @@ public class CustomMentoringRepositoryImpl implements CustomMentoringRepository
+        List<Long> categoryIds
+    ) {
+        // 결과 행을 튜플 타입(Mentoring, MentoringStatistics)으로 가져온다.
+        List<Tuple> rows = jpaQueryFactory
```
p3: 좀 더 의미가 드러나는 변수명은 어떨까요? ex `mentoringAndStatistics`

kimdevv replied:
그렇겠네요. 수정하겠습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/MentoringController.java:
```diff
@@ -50,21 +50,6 @@ public ResponseEntity<Void> registerMentoring(
                 .build();
     }
 
-    @GetMapping("/mentorings")
-    public ResponseEntity<List<MentoringSummaryResponse>> getMentoringSummaries(
-            @RequestParam(required = false) String categoryTitle1,
-            @RequestParam(required = false) String categoryTitle2,
-            @RequestParam(required = false) String categoryTitle3
-    ) {
-        List<MentoringSummaryResponse> responseBody = mentoringService.findMentoringSummaries(
-                categoryTitle1,
-                categoryTitle2,
-                categoryTitle3
-        );
-        return ResponseEntity.status(HttpStatus.OK)
-                .body(responseBody);
-    }
-
```
p3 : 결국 deprecated 되었군요

kimdevv replied:
그렇게 되었습니다. 더이상 사용하지 않는 API라서 제거했습니다.

jye326 reviewed (approved):
수고하셨습니다~ 코멘트 확인해주세요!

kimdevv committed:
refactor: MentoringStatistics의 id를 mentoringId로 변경

- MentoringStatistics에서 필드 이름을 id에서 mentoringId로 수정
- 관련 메서드와 쿼리에서 id를 mentoringId로 변경하여 일관성 유지

kimdevv committed:
refactor: 변수명 명확성 개선 및 관련 코드 수정

- rows를 mentoringAndMentoringStatistics로 변수명 변경
- 관련 코드에서 변수명 수정으로 가독성과 명확성 강화

kimdevv committed:
Merge branch 'develop' into feature/778-order-most-reservation

# Conflicts:
#	backend/fittoring/src/test/java/fittoring/mentoring/business/service/AuthServiceTest.java
#	backend/fittoring/src/test/java/fittoring/mentoring/business/service/CertificateServiceTest.java
#	backend/fittoring/src/test/java/fittoring/mentoring/business/service/MemberServiceTest.java

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -159,9 +158,20 @@ private MentoringResponse getMentoringWithRelations(Mentoring mentoring) {
-                .orElseGet(() -> RatingStatsDto.defaultOf(mentoringId));
+        MentoringStatistics mentoringStatistics = mentoringStatisticsRepository.findById(mentoringId)
+            .orElseThrow(() -> new MentoringNotFoundException(BusinessErrorMessage.MENTORING_NOT_FOUND.getMessage()));
+        System.out.println(mentoringStatistics.getReviewCount() + " " + mentoringStatistics.getRatingSum() + " ===========================================================");
```
p2: 출력문은 지워져야 할 것 같네요!

kimdevv replied:
헉 이게 있었군요 감사합니다.

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -159,9 +158,20 @@ private MentoringResponse getMentoringWithRelations(Mentoring mentoring) {
     private RatingStatsDto getRatingStatsDto(Long mentoringId) {
-        return reviewRepository.findRatingStatsByMentoringId(mentoringId)
-                .orElseGet(() -> RatingStatsDto.defaultOf(mentoringId));
+        MentoringStatistics mentoringStatistics = mentoringStatisticsRepository.findById(mentoringId)
+            .orElseThrow(() -> new MentoringNotFoundException(BusinessErrorMessage.MENTORING_NOT_FOUND.getMessage()));
```
p3: 조회 로직도 통계 테이블을 사용하도록 변경되었군요 👍

ppparkta reviewed (approved):
좋습니다. 코멘트 몇 개 남겼습니다. 확인 부탁드려요!
멘토링 개설 수정중인데 머지받기가 무섭네요 ㅎㅎ...

kimdevv committed:
refactor: 디버그용 출력문 제거

- getRatingStatsDto 메서드에서 불필요한 System.out.println 제거
- 코드 가독성 및 유지보수성 강화

kimdevv committed:
refactor: QueryDSL 관련 클래스의 패키지 수정

kimdevv committed:
feat: CustomMentoringRepository 패키지 원상복구
